### ISSUE TITLE
deprecated Yeoman generator#invoke called from loopback:model on property creation

### ISSUE THREAD

**Issue by Sequoia** [strongloop/generator-loopback#116]
```
Enter an empty property name when done.
? Property name: name
(!) generator#invoke() is deprecated. Use generator#composeWith() - see http://yeoman.io/authoring/composability.html
   invoke   loopback:property
? Property type: (Use arrow keys)
❯ string
  number
  boolean
  object
  array
  date
  buffer
(Move up and down to reveal more choices)
```
not clear if #109 fixes this, I don't see references to `invoke` in the patch.

**Comment by Sequoia** [comment]
Using the scaffolder is one of the first interactions new users have with LoopBack; this could def. throw someone off track.

**Comment by bajtos** [comment]
This is tricky. We are using yeoman generators in a way that was supported in pre-v0.17 days, but that is different from the new direction where yeoman is heading nowadays.

Our design is based on assumption that generators work like javascript methods: if we want to compose "loopback:model" with "loopback:property", we can simply `generator#invoke()` the later and the execution of the former would be postponed until the nested generator finishes.

However, that's no longer the case. These days, Yeoman uses a run queue with named groups (see "The Run Loop" in [docs](http://yeoman.io/authoring/running-context.html)). When two generators are composed, their steps/actions are interleaved together via these named groups (e.g. "prompting", "configuring", "writing", "install", etc.). This composition does not allow repeated invocation of a nested generator, as we are doing now.

I see two ways how to fix the problem short-term: either pin the yeoman-generator dependency to the version that does not report the deprecation warning, or remove invocation of "loopback:property" from "loopback:model" - users will have to manually and repeatedly call "loopback:property" to add all properties they need.

For long-term, I think we should rework generator-loopback, refactor most of the prompts & actions so that they don't depend on yeoman features and the yeoman integration becomes just a thin wrapper. As a side effect, it should become easier to write unit-tests for prompts to verify input validation for example.

Mockup for illustration:

````js
// model generator
generators.Base.extend({
  prompting: {
     promptForModelData: function() { 
       this.modelData = prompts.modelData();
     },
     promptForProperties: function() {
       this.properties = [];
       for (;;) {
         var prop = prompts.propertyData(this.modelData.name);
         if (!prop.name) break;
         this.properties.push(prop);
      }
  },
  writing: {
     createModel: function() {
       scaffold.model(this.modelData);
     },
     createProperties: function() {
       this.properties.forEach(function(prop) {
          scaffold.property(prop);
       });
     }
  }
});
```

/cc @kraman @ritch @strongloop/loopback-dev thoughts?

**Comment by kraman** [comment]
@bajtos I vote for pinning the version while we work on a better fix or move away from loopback yoman generator

**Comment by jannyHou** [comment]
I will figure out the version that doesn't give warning ASAP. I think no matter which short term workaround we take, It's always better to know the "safe" version.

**Comment by bajtos** [comment]
> I will figure out the version that doesn't give warning ASAP. 

@jannyHou the `package.json` says ` "yeoman-generator": "^0.21.1",`, I would start with looking at the version `0.21.1`.

**Comment by jannyHou** [comment]
Thank you @bajtos the latest available version is 0.18.10 but it causes some test failures, I am looking into those failures now :) Details:
https://github.com/strongloop-internal/scrum-loopback/issues/754#issuecomment-175231847

**Comment by bajtos** [comment]
I created a follow-up issue for the long-term solution here: https://github.com/strongloop/generator-loopback/issues/139
